File: apps/sdkwork-rtc-h5/packages/sdkwork-rtc-h5-admin-core/src/services/providerSchemaService.ts
import type { AuthTokenManager } from "@sdkwork/sdk-common";

import type { ProviderConfigSchema } from "../types/providerSchema";
import { readSdkWorkItem } from "../sdk/index.js";
import { resolveBackendRtcClient, type RtcBackendClientOptions, type RtcBackendClientSource } from "./backendClient";

export class ProviderSchemaService {
  private readonly client;

  constructor(
    baseUrlOrClient: RtcBackendClientSource,
    tokenManagerOrOptions?: AuthTokenManager | RtcBackendClientOptions,
  ) {
    this.client = resolveBackendRtcClient(baseUrlOrClient, tokenManagerOrOptions);
  }

  async listSchemas(): Promise<ProviderConfigSchema[]> {
    const response = await this.client.rtcProviderSchemas.rtc.providerSchemas.list();
    if (!response.data) {
      throw new Error("Invalid response: missing provider schema data");
    }
    return readSdkWorkItem<ProviderConfigSchema[]>(response.data);
  }

  async getSchema(provider: string): Promise<ProviderConfigSchema> {
    const response = await this.client.rtcProviderSchemas.rtc.providerSchemas.retrieve(provider);
    if (!response.data) {
      throw new Error(`RTC provider schema not found: ${provider}`);
    }
    return readSdkWorkItem<ProviderConfigSchema>(response.data);
  }
}

File: apps/sdkwork-rtc-flutter-mobile/packages/sdkwork_rtc_flutter_mobile_admin_core/lib/src/services/provider_schema_service.dart
import '../admin_sdk_mapper.dart';
import '../backend_rtc_client.dart';
import '../models/provider_schema.dart';

class ProviderSchemaService {
  final SdkworkBackendClient _client;

  ProviderSchemaService(this._client);

  Future<List<ProviderConfigSchema>> listSchemas() async {
    final response = await _client.rtcProviderSchemas.list();
    final data = backendResponseData(response);
    final items = data?['items'] ?? data;
    if (items is List<dynamic>) {
      return items
          .whereType<Map<String, dynamic>>()
          .map(ProviderConfigSchema.fromJson)
          .toList();
    }
    if (data is Map<String, dynamic>) {
      return [ProviderConfigSchema.fromJson(data)];
    }
    return [];
  }

  Future<ProviderConfigSchema?> getSchema(String provider) async {
    final response = await _client.rtcProviderSchemas.retrieve(provider);
    final data = backendResponseData(response);
    if (data == null) {
      return null;
    }
    return ProviderConfigSchema.fromJson(data);
  }
}
